Deep sleep not compiling on ESP32-C6 with 2025.8.0b1 and dev.

### The problem

Previously deep sleep was working in dev for an ESP32-C6. It now comes up with the errors below.

Suspect it maybe related to PR #9395 

### Which version of ESPHome has the issue?

2025.8.0-b1

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

deep_sleep

### YAML Config

```yaml
esphome:
  name: moisture-sensor-1
  friendly_name: Moisture sensor 1
  on_boot: 
    then:
      - switch.turn_on: moisture_sensor_power

esp32:
  board: esp32-c6-devkitm-1
  framework:
    type: esp-idf




deep_sleep:
  id: moisture_sensor_1
  run_duration: 25sec
  sleep_duration: 4hours
```

### Anything in the logs that might be useful for us?

```txt
src/esphome/components/deep_sleep/deep_sleep_esp32.cpp: In member function 'void esphome::deep_sleep::DeepSleepComponent::deep_sleep_()':
src/esphome/components/deep_sleep/deep_sleep_esp32.cpp:123:5: error: 'gpio_deep_sleep_hold_en' was not declared in this scope; did you mean 'gpio_sleep_sel_en'?
  123 |     gpio_deep_sleep_hold_en();
      |     ^~~~~~~~~~~~~~~~~~~~~~~
      |     gpio_sleep_sel_en
*** [.pioenvs/moisture-sensor-1/src/esphome/components/deep_sleep/deep_sleep_esp32.cpp.o] Error 1
========================== [FAILED] Took 6.88 seconds ==========================
```

### Additional information

_No response_

--- a/esphome/components/deep_sleep/deep_sleep_esp32.cpp
+++ b/esphome/components/deep_sleep/deep_sleep_esp32.cpp
@@ -83,7 +83,11 @@ void DeepSleepComponent::deep_sleep_() {
     }
     gpio_sleep_set_direction(gpio_pin, GPIO_MODE_INPUT);
     gpio_hold_en(gpio_pin);
+#if !defined(USE_ESP32_VARIANT_ESP32C6)
+    // ESP32-C6 doesn't have gpio_deep_sleep_hold_en() since it has SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
+    // For C6, gpio_hold_en() is sufficient to hold the pin state during deep sleep
     gpio_deep_sleep_hold_en();
+#endif
     bool level = !this->wakeup_pin_->is_inverted();
     if (this->wakeup_pin_mode_ == WAKEUP_PIN_MODE_INVERT_WAKEUP && this->wakeup_pin_->digital_read()) {
       level = !level;
@@ -120,7 +124,11 @@ void DeepSleepComponent::deep_sleep_() {
     }
     gpio_sleep_set_direction(gpio_pin, GPIO_MODE_INPUT);
     gpio_hold_en(gpio_pin);
+#if !defined(USE_ESP32_VARIANT_ESP32C6)
+    // ESP32-C6 doesn't have gpio_deep_sleep_hold_en() since it has SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
+    // For C6, gpio_hold_en() is sufficient to hold the pin state during deep sleep
     gpio_deep_sleep_hold_en();
+#endif
     bool level = !this->wakeup_pin_->is_inverted();
     if (this->wakeup_pin_mode_ == WAKEUP_PIN_MODE_INVERT_WAKEUP && this->wakeup_pin_->digital_read()) {
       level = !level;

--- a/esphome/components/deep_sleep/deep_sleep_esp32.cpp
+++ b/esphome/components/deep_sleep/deep_sleep_esp32.cpp
@@ -1,4 +1,5 @@
 #ifdef USE_ESP32
+#include "soc/soc_caps.h"
 #include "driver/gpio.h"
 #include "deep_sleep_component.h"
 #include "esphome/core/log.h"
@@ -83,9 +84,9 @@ void DeepSleepComponent::deep_sleep_() {
     }
     gpio_sleep_set_direction(gpio_pin, GPIO_MODE_INPUT);
     gpio_hold_en(gpio_pin);
-#if !defined(USE_ESP32_VARIANT_ESP32C6)
-    // ESP32-C6 doesn't have gpio_deep_sleep_hold_en() since it has SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
-    // For C6, gpio_hold_en() is sufficient to hold the pin state during deep sleep
+#if !SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
+    // Some ESP32 variants support holding a single GPIO during deep sleep without this function
+    // For those variants, gpio_hold_en() is sufficient to hold the pin state during deep sleep
     gpio_deep_sleep_hold_en();
 #endif
     bool level = !this->wakeup_pin_->is_inverted();
@@ -124,9 +125,9 @@ void DeepSleepComponent::deep_sleep_() {
     }
     gpio_sleep_set_direction(gpio_pin, GPIO_MODE_INPUT);
     gpio_hold_en(gpio_pin);
-#if !defined(USE_ESP32_VARIANT_ESP32C6)
-    // ESP32-C6 doesn't have gpio_deep_sleep_hold_en() since it has SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
-    // For C6, gpio_hold_en() is sufficient to hold the pin state during deep sleep
+#if !SOC_GPIO_SUPPORT_HOLD_SINGLE_IO_IN_DSLP
+    // Some ESP32 variants support holding a single GPIO during deep sleep without this function
+    // For those variants, gpio_hold_en() is sufficient to hold the pin state during deep sleep
     gpio_deep_sleep_hold_en();
 #endif
     bool level = !this->wakeup_pin_->is_inverted();

--- a/tests/components/deep_sleep/test.esp32-c6-idf.yaml
+++ b/tests/components/deep_sleep/test.esp32-c6-idf.yaml
@@ -1,0 +1,5 @@
+substitutions:
+  wakeup_pin: GPIO4
+
+<<: !include common.yaml
+<<: !include common-esp32.yaml

--- a/tests/components/deep_sleep/test.esp32-s2-idf.yaml
+++ b/tests/components/deep_sleep/test.esp32-s2-idf.yaml
@@ -1,0 +1,5 @@
+substitutions:
+  wakeup_pin: GPIO4
+
+<<: !include common.yaml
+<<: !include common-esp32.yaml

--- a/tests/components/deep_sleep/test.esp32-s3-idf.yaml
+++ b/tests/components/deep_sleep/test.esp32-s3-idf.yaml
@@ -1,0 +1,5 @@
+substitutions:
+  wakeup_pin: GPIO4
+
+<<: !include common.yaml
+<<: !include common-esp32.yaml

